### ISSUE TITLE
media latching with asymmetric RTP.

### ISSUE THREAD

**Issue by patrickarton** [sipwise/rtpengine#1566]
Hello i want to enable asymetric RTP with media latching but  that seems impossible.

Basically Device1  uses **IP A** and **Port a** to receive RTP (ie SDP cline=**A** and mline[RTP port]=**a**) and it uses **IP A** and **Port b** to transmit or **IP B** and **port b** to transmit.

RTPEngine must forward RTP it receives from the remote party to **address "A" and port "a"** for
Device1.In other words, RTPEngine **must always transmit** RTP stream to the **IP address and RTP port received in SDP** offer/answer.

Now when Device1 uses **IP A** and **Port b** or **IP B** and **port b** to transmit RTPEngine must latch to that IP and port
when it receives First RTP. (unfortunately STRICT_SOURCE flag does not work).

So basically what i want is : 

- RTPEngine must always transmit RTP stream to the IP address and RTP port received in SDP offer/answer
- for asymmetric RTP latch to first IP and port from which rtp is sent and drop if any change.

**Comment by patrickarton** [comment]
> Passing the `asymmetric` flag in the offer/answer flags should do the trick.

yes but i also want the **STRICT_SOURCE** flag to apply to it but it does not work. it will fail on the first check. it will not let the first asymmetric rtp

```
	/* do not pay attention to source addresses of incoming packets for asymmetric streams */
	if (MEDIA_ISSET(phc->mp.media, ASYMMETRIC) || rtpe_config.endpoint_learning == EL_OFF)
		PS_SET(phc->mp.stream, CONFIRMED);

	/* confirm sinks for unidirectional streams in order to kernelize */
	if (MEDIA_ISSET(phc->mp.media, UNIDIRECTIONAL)) {
		for (GList *l = phc->sinks->head; l; l = l->next) {
			struct sink_handler *sh = l->data;
			PS_SET(sh->sink, CONFIRMED);
		}
	}

	/* if we have already updated the endpoint in the past ... */
	if (PS_ISSET(phc->mp.stream, CONFIRMED)) {
		/* see if we need to compare the source address with the known endpoint */
		if (PS_ISSET2(phc->mp.stream, STRICT_SOURCE, MEDIA_HANDOVER)) {
			endpoint = phc->mp.fsin;
			mutex_lock(&phc->mp.stream->out_lock);

			int tmp = memcmp(&endpoint, &phc->mp.stream->endpoint, sizeof(endpoint));
			if (tmp && PS_ISSET(phc->mp.stream, MEDIA_HANDOVER)) {
				/* out_lock remains locked */
				ilog(LOG_INFO | LOG_FLAG_LIMIT, "Peer address changed to %s%s%s",
						FMT_M(endpoint_print_buf(&phc->mp.fsin)));
				phc->unkernelize = 1;
				phc->unconfirm = 1;
				phc->update = 1;
				phc->mp.stream->endpoint = phc->mp.fsin;
				goto update_addr;
			}

			mutex_unlock(&phc->mp.stream->out_lock);

			if (tmp && PS_ISSET(phc->mp.stream, STRICT_SOURCE)) {
				ilog(LOG_INFO | LOG_FLAG_LIMIT, "Drop due to strict-source attribute; "
						"got %s%s:%d%s, "
						"expected %s%s:%d%s",
					FMT_M(sockaddr_print_buf(&endpoint.address), endpoint.port),
					FMT_M(sockaddr_print_buf(&phc->mp.stream->endpoint.address),
					phc->mp.stream->endpoint.port));
				atomic64_inc(&phc->mp.stream->stats.errors);
				ret = -1;
			}
		}
		phc->kernelize = 1;
		goto out;
	}
```

**Comment by rfuchs** [comment]
So you want to keep endpoint learning enabled, but only for the purpose of the sender address/port, but not for determining the reverse destination address/port?

**Comment by patrickarton** [comment]
> So you want to keep endpoint learning enabled, but only for the purpose of verifying the sender address/port, but not for determining the reverse destination address/port?

yes exactly i want device to be able to use asymmetric RTP. and on rtpengine side i want also to be able to latch on that learning endpoint (with STRICT_SOURCE ) as soon as i get first packet.

**Comment by rfuchs** [comment]
Ok, that's not supported then. Right now it's either/or. Patch welcome.
